OrbitBase/include/OrbitBase/Action.h

```diff
@@ -21,7 +21,8 @@ class Action {
 template <typename F>
 class NullaryFunctorAction : public Action {
  public:
-  explicit NullaryFunctorAction(F functor) : functor_(functor) {}
+  explicit NullaryFunctorAction(F&& functor)
+      : functor_(std::forward<F>(functor)) {}
```

Comment:
Universal forwarding only works here when the constructor is templated independently from the class.
```
template<typename T>
explicit NullaryFunctionAction(T&& functor) : functor_(std::forward<T>(functor)) {}
```

BTW, `CreateAction` has the same problem.

Reply:
I will fix it a separate PR, this like only does clang-format I think.

Reply:
Are you sure CreateAction has the same problem? seems like there is type deduction is going on - so it should be universal reference? It is a standalone function not part of the class...

Reply:
Ok, let's say the problem is similar. `CreateAction` needs to remove the reference from F:
`return std::make_unique<NullaryFunctorAction<std::remove_reference_t<F>>>(std::forward<F>(functor));`

Otherwise the action might only store a reference.

Reply:
It's not only `clang-format`, I added `std::forward` here, otherwise the copy constructor of `functor_` would always be called (which is a compilation error when `F` is only movable). So from this point of view it seems that universal forwarding is at least partially working here, or what am I missing? I only understand so much of all this stuff.

Reply:
By the way, the issue happened with a lambda that captured a move-only-type variable (in particular a `unique_ptr`) by value. Let's make sure to fix this as it's in `OrbitBase`.

Reply:
> `template<typename T>`
> `explicit NullaryFunctionAction(T&& functor) : functor_(std::forward<T>(functor)) {}`

This gives the following clang-tidy:
"Constructor accepting a forwarding reference can hide the copy and move constructors."

Are we ok with it?

Reply:
> So what's happening here: By adding `&&` it gets an r-value reference, but it won't be a univeral reference when the template parameter can't be deduced freely by the compiler. So I think it's working in that case because you pass in a r-value reference. But passing in a l-value-reference won't work anymore, so it's not universal.

Yep, makes sense, thanks.

> The clang-tidy-warning is legit. We have to take care of that. Does defaulting the copy- and move-constructors help?

It doesn't. But I've been playing around with `std::enable_if`, please have a look.

Reply:
I was hoping this could be done without SFINAE. :see_no_evil: Sorry for that. Your enable_if looks good.

Reply:
Is it bad it's done with SFINAE? I like some confusing template magic here and there. 😄